### PR TITLE
Fix view change incorrect node ejection, add random DS candidate leader selection and macros for test scenarios

### PR DESCRIPTION
**Pull request by bb111189**
## Description

This PR fixes 
- issue https://github.com/Zilliqa/Issues/issues/212 where view change will inaccurate eject wrong DS node to the end of the DS committee. This bug was a result of the newly implemented random DS leader selection. 
- re-election of same DS leader during multiple view change 

In additional, this PR implements the following
- Random DS leader selection for view change
- Re computation of candidate leader if it happen to be the same faulty leader
- Revamping VC block header
-- Removal of `expectedCandidateLeaderIndex`
-- Addition of `faultyLeaders` vector
- Add Marco for view change testing
- Revamps the logic for updating the DS composition 
- Removal of termination test as it is covered by stalling test

### Implementation

- [x] **ready for review**
- [x] **Re-test after new changes (from PR reviews)**

### Integration Test (Core Team)

Single consensus failure
- [x] [vc1] DS leader stalls at DS block consensus 
- [x] [vc3] DS leader stalls at Final block consensus 

Multiple failure (after view change is completed) 
- [x] [vc2] DS leader stalls at DS block consensus and 2 candidate leaders stall at DS block consensus 
- [x] [vc4] DS leader stalls at Final block consensus and 2 candidate leaders stall at Final block consensus 

Multiple failure (with view change consensus failure) 
- [x] [vc1 vc5] DS leader stalls at DS block consensus and candidate leaders stall at View Change consensus
- [x] [vc3 vc5] DS leader stalls at Final block consensus and 2 candidate leaders stall at View Change consensus
- [x] [vc1 vc6] DS leader stalls at DS block consensus and candidate leaders stall at View Change consensus
- [x] [vc3 vc6] DS leader stalls at Final block consensus and 2 candidate leaders stall at View Change consensus

Other
- [x] Medium-scale cloud test (functionality test. Not view change test) (72c832b4beeed225e77deb7b467e55b53ea0fef0)

**Re-test**
Single consensus failure
- [x] [vc1] DS leader stalls at DS block consensus 
- [x] [vc3] DS leader stalls at Final block consensus 

Multiple failure (after view change is completed) 
- [x] [vc2] DS leader stalls at DS block consensus and 2 candidate leaders stall at DS block consensus 
- [x] [vc4] DS leader stalls at Final block consensus and 2 candidate leaders stall at Final block consensus 

Multiple failure (with view change consensus failure) 
- [x] [vc1 vc5] DS leader stalls at DS block consensus and candidate leaders stall at View Change consensus
- [x] [vc3 vc5] DS leader stalls at Final block consensus and 2 candidate leaders stall at View Change consensus
- [x] [vc1 vc6] DS leader stalls at DS block consensus and candidate leaders stall at View Change consensus
- [x] [vc3 vc6] DS leader stalls at Final block consensus and 2 candidate leaders stall at View Change consensus

### CODE REVIEW

File: src/libNode/ViewChangeBlockProcessing.cpp [hunk 1]
```diff
@@ -238,14 +210,20 @@ bool Node::ProcessVCBlockCore(const VCBlock& vcblock) {
...
   }
 
-  {
-    lock_guard<mutex> g(m_mediator.m_mutexDSCommittee);
-    for (unsigned int x = 0; x < newCandidateLeader; x++) {
-      // TODO: If VC select a random
-      // leader, we need to change the way
-      // we update ds composition.
-      UpdateDSCommiteeCompositionAfterVC(*m_mediator.m_DSCommittee);
+  for (const auto& faultyLeader : vcblock.GetHeader().GetFaultyLeaders()) {
```
Comment (KaustubhShamshery) [inline src/libNode/ViewChangeBlockProcessing.cpp]: Hey, can you make this a separate function as I need it in the Directory Block verification

Comment (bb111189) [reply src/libNode/ViewChangeBlockProcessing.cpp]: Will extract to a new function then

Comment (KaustubhShamshery) [reply src/libNode/ViewChangeBlockProcessing.cpp]: Thanks

Comment (bb111189) [reply src/libNode/ViewChangeBlockProcessing.cpp]: Implemented

File: src/libNode/ViewChangeBlockProcessing.cpp [hunk 2]
```diff
@@ -238,19 +210,29 @@ bool Node::ProcessVCBlockCore(const VCBlock& vcblock) {
     return false;
   }
```
Comment (KaustubhShamshery) [inline src/libNode/ViewChangeBlockProcessing.cpp]: I think you may need the mutex for DS comm. Also there is deperecared UpdateDSCommiteeCompositionAfterVC(deque ..) I see.

File: src/libNode/ViewChangeBlockProcessing.cpp [hunk 3]
```diff
@@ -238,19 +210,29 @@ bool Node::ProcessVCBlockCore(const VCBlock& vcblock) {
...
-      UpdateDSCommiteeCompositionAfterVC(*m_mediator.m_DSCommittee);
-    }
-  }
+  UpdateDSCommiteeCompositionAfterVC(vcblock);
 
   return true;
 }
 
+void Node::UpdateDSCommiteeCompositionAfterVC(const VCBlock& vcblock) {
+  for (const auto& faultyLeader : vcblock.GetHeader().GetFaultyLeaders()) {
```
[the remark this answers is not in the document]
Comment (bb111189) [reply src/libNode/ViewChangeBlockProcessing.cpp]: Why `deque`? Faulty leaders is inside a vector in vcblock header

File: src/libDirectoryService/ViewChangePreProcessing.cpp [hunk 4]
```diff
@@ -294,17 +367,90 @@ bool DirectoryService::ComputeNewCandidateLeader() {
...
+
+  if (latestBlockType == BlockType::VC) {
+    if (!BlockStorage::GetBlockStorage().GetVCBlock(
+            get<BlockLinkIndex::BLOCKHASH>(bl), prevVCBlockptr)) {
+      LOG_GENERAL(WARNING, "could not get vc block "
+                               << get<BlockLinkIndex::BLOCKHASH>(bl));
+      return false;
+    }
+
+    if (prevVCBlockptr->GetHeader().GetViewChangeEpochNo() !=
```
Comment (kaikawaliu) [inline src/libDirectoryService/ViewChangePreProcessing.cpp]: Should it be `GetViewChangeDSEpochNo()`?

File: src/libDirectoryService/ViewChangePreProcessing.cpp [hunk 5]
```diff
@@ -294,17 +367,90 @@ bool DirectoryService::ComputeNewCandidateLeader() {
...
+
+  vector<unsigned char> vcCounterBytes;
+  Serializable::SetNumber<uint32_t>(vcCounterBytes, 0, m_viewChangeCounter,
+                                    sizeof(uint32_t));
+  sha2.Update(vcCounterBytes);
+  uint16_t lastBlockHash = DataConversion::charArrTo16Bits(sha2.Finalize());
+  uint32_t candidateLeaderIndex =
+      lastBlockHash % (m_mediator.m_DSCommittee->size());
+
+  while (candidateLeaderIndex == m_consensusLeaderID) {
```
Comment (kaikawaliu) [inline src/libDirectoryService/ViewChangePreProcessing.cpp]: what if the calculated leader index also points to one of the faulty leaderS that hve been moved to the back of the `m_DSCommittee`? Maybe the `candidateLeaderIndex = lastBlockHash % (mediator.m_DSCommittee->size() - prevVCBlockPtr->GetHeader().GetFaultyLeaders().size())` can help?

File: src/libDirectoryService/ViewChangePreProcessing.cpp [hunk 6]
```diff
@@ -266,15 +333,21 @@ bool DirectoryService::ComputeNewCandidateLeader() {
...
+      "Composing new vc block with vc count at "
+          << m_viewChangeCounter << " and candidate leader is at index "
+          << candidateLeaderIndex << ". "
+          << m_mediator.m_DSCommittee->at(candidateLeaderIndex).second << " "
+          << DataConversion::SerializableToHexStr(
+                 m_mediator.m_DSCommittee->at(candidateLeaderIndex).first));
 
   Peer newLeaderNetworkInfo;
-  if (m_mediator.m_DSCommittee->at(m_viewChangeCounter).second == Peer()) {
+  if (m_mediator.m_DSCommittee->at(candidateLeaderIndex).second == Peer()) {
```
Comment (KaustubhShamshery) [inline src/libDirectoryService/ViewChangePreProcessing.cpp]: I suggest also check selfKey here

File: src/libDirectoryService/ViewChangePreProcessing.cpp (hunk 4 above)
Comment (bb111189) [reply src/libDirectoryService/ViewChangePreProcessing.cpp]: Fixed

File: src/libNode/ViewChangeBlockProcessing.cpp (hunk 3 above)
[the remark this answers is not in the document]
Comment (bb111189) [reply src/libNode/ViewChangeBlockProcessing.cpp]: implemented

File: CMakeLists.txt [hunk 7]
```diff
@@ -58,18 +57,54 @@ if(OPENCL_MINE)
 endif()
 
 if(CUDA_MINE)
+    message(STATUS "Found cuda")
```
Comment (ShengguangXiao) [inline CMakeLists.txt]: Should it change to message(STATUS "CUDA enabled")?

File: src/libDirectoryService/ViewChangePostProcessing.cpp [hunk 8]
```diff
@@ -90,135 +93,127 @@ void DirectoryService::ProcessViewChangeConsensusWhenDone() {
...
-          LOG_GENERAL(INFO, "New m_consensusMyID  is " << m_consensusMyID);
-        }
+      // Remove faulty leader from the current ds committee structure temporary
+      if (iterFaultyLeader != m_mediator.m_DSCommittee->end()) {
+        m_mediator.m_DSCommittee->erase(iterFaultyLeader);
+      } else {
+        LOG_GENERAL(WARNING, "Cannot find "
+                                 << faultyLeader.second
+                                 << " to eject to back of ds committee");
+        // TODO: Handle this situation. This siutation shouldn't be
```
Comment (ansnunez) [inline src/libDirectoryService/ViewChangePostProcessing.cpp]: Perhaps for scenarios like this we should start adding FATAL logs, so it's easier to discover in the testnet logs

File: src/libDirectoryService/ViewChangePostProcessing.cpp [hunk 9]
```diff
@@ -90,135 +93,127 @@ void DirectoryService::ProcessViewChangeConsensusWhenDone() {
...
+             selfPubKPeerPair);
+
+    if (iterConsensusMyID != m_mediator.m_DSCommittee->end()) {
+      m_consensusMyID =
+          distance(m_mediator.m_DSCommittee->begin(), iterConsensusMyID);
     } else {
-      m_mediator.m_node->m_isPrimary = false;
-      LOG_EPOCH(INFO, to_string(m_mediator.m_currentEpochNum).c_str(),
-                "I am backup member of the DS shard");
+      LOG_GENERAL(WARNING,
```
Comment (ansnunez) [inline src/libDirectoryService/ViewChangePostProcessing.cpp]: Similar comment here, maybe we can upgrade this to FATAL

File: src/libDirectoryService/ViewChangePostProcessing.cpp [hunk 10]
```diff
@@ -90,135 +93,127 @@ void DirectoryService::ProcessViewChangeConsensusWhenDone() {
...
+
+  if (iterConsensusLeaderID != m_mediator.m_DSCommittee->end()) {
+    m_consensusLeaderID =
+        distance(m_mediator.m_DSCommittee->begin(), iterConsensusLeaderID);
   } else {
-    LOG_EPOCH(WARNING, to_string(m_mediator.m_currentEpochNum).c_str(),
-              "View change completed but it seems wrong to me."
-                  << "expectedLeader: " << expectedLeader
-                  << "newLeaderNetworkInfo: " << newLeaderNetworkInfo);
+    LOG_GENERAL(WARNING, "Cannot find new leader in the ds committee");
```
Comment (ansnunez) [inline src/libDirectoryService/ViewChangePostProcessing.cpp]: Similar comment here, maybe we can upgrade this to FATAL

File: src/libDirectoryService/ViewChangePreProcessing.cpp [hunk 11]
```diff
@@ -294,17 +378,89 @@ bool DirectoryService::ComputeNewCandidateLeader() {
...
+                "Computed candidate leader is current faulty ds leader. Index: "
+                    << candidateLeaderIndex);
+    sha2.Update(sha2.Finalize());
+    lastBlockHash = DataConversion::charArrTo16Bits(sha2.Finalize());
+    candidateLeaderIndex = lastBlockHash % (m_mediator.m_DSCommittee->size());
+    LOG_GENERAL(INFO, "Re-computed candidate leader is at index: "
+                          << candidateLeaderIndex
+                          << " VC counter: " << m_viewChangeCounter);
+  }
+  sha2.Reset();
```
Comment (ansnunez) [inline src/libDirectoryService/ViewChangePreProcessing.cpp]: Is this line needed?

File: src/libDirectoryService/ViewChangePreProcessing.cpp [hunk 12]
```diff
@@ -95,15 +95,70 @@ bool DirectoryService::ViewChangeValidator(
     return false;
   }
 
-  if (m_mediator.m_DSCommittee->at(m_viewChangeCounter).second !=
+  // Verify candidate leader index
+  uint32_t candidateLeaderIndex = CalculateNewLeaderIndex();
```
Comment (ansnunez) [inline src/libDirectoryService/ViewChangePreProcessing.cpp]: The other part where `CalculateNewLeaderIndex` is called doesn't seem to have the same verification code block below. Do we need to add it there? Actually here it seems unnecessary as this if condition should never happen (or we can keep it but change to FATAL).

File: src/libNode/ViewChangeBlockProcessing.cpp (hunk 2 above)
Comment (bb111189) [reply src/libNode/ViewChangeBlockProcessing.cpp]: implemented

File: src/libDirectoryService/ViewChangePreProcessing.cpp (hunk 12 above)
Comment (bb111189) [reply src/libDirectoryService/ViewChangePreProcessing.cpp]: Changed to FATAL

**Comment by bb111189** [comment]
Addressed all feedbacks. Will do a re-test of all the scenarios again.

File: src/libNode/ViewChangeBlockProcessing.cpp [hunk 13]
```diff
@@ -238,19 +204,29 @@ bool Node::ProcessVCBlockCore(const VCBlock& vcblock) {
...
+    const VCBlock& vcblock, deque<pair<PubKey, Peer>>& dsComm) {
+  for (const auto& faultyLeader : vcblock.GetHeader().GetFaultyLeaders()) {
+    deque<pair<PubKey, Peer>>::iterator it =
+        find(dsComm.begin(), dsComm.end(), faultyLeader);
+
+    // Remove faulty leader from the current
+    if (it != dsComm.end()) {
+      dsComm.erase(it);
+    } else {
+      LOG_GENERAL(WARNING, "Cannot find the ds leader to eject");
```
Comment (ansnunez) [inline src/libNode/ViewChangeBlockProcessing.cpp]: If this condition is also impossible to hit, we can change to FATAL now

Comment (bb111189) [reply src/libNode/ViewChangeBlockProcessing.cpp]: Changed to FATAL

**Pull request merged**